File: Problemset/1749-Maximum Absolute Sum of Any Subarray/Prefix Sum.rb
# @param {Integer[]} nums
# @return {Integer}
def max_absolute_sum(nums)
  sum = 0
  max_sum = 0
  min_sum = 0
  ret = 0

  nums.each do |x|
    sum += x
    max_sum = [max_sum, sum].max
    min_sum = [min_sum, sum].min
    ret = [ret, (sum - max_sum).abs, (sum - min_sum).abs].max
  end

  ret
end

File: Problemset/1749-Maximum Absolute Sum of Any Subarray/Prefix Sum.rs
impl Solution {
    pub fn max_absolute_sum(nums: Vec<i32>) -> i32 {
        let mut sum = 0;
        let mut max_sum = 0;
        let mut min_sum = 0;
        let mut ret = 0;

        for x in nums {
            sum += x;
            max_sum = max_sum.max(sum);
            min_sum = min_sum.min(sum);
            ret = ret.max((sum - max_sum).abs()).max((sum - min_sum).abs());
        }

        ret
    }
}
